lib/smack/DSAWrapper.cpp

```diff
@@ -22,185 +23,96 @@ RegisterPass<DSAWrapper>
 
-std::vector<const llvm::DSNode *>
-DSAWrapper::collectStaticInits(llvm::Module &M) {
-  std::vector<const llvm::DSNode *> sis;
+void DSAWrapper::collectStaticInits(llvm::Module &M) {
   for (GlobalVariable &GV : M.globals()) {
     if (GV.hasInitializer()) {
       if (auto *N = getNode(&GV)) {
-        sis.push_back(N);
+        staticInits.insert(N);
       }
     }
   }
-  return sis;
-}
-
-DSGraph *DSAWrapper::getGraphForValue(const Value *V) {
-  if (const Instruction *I = dyn_cast<Instruction>(V))
-    return TD->getDSGraph(*I->getParent()->getParent());
-  else if (const Argument *A = dyn_cast<Argument>(V))
-    return TD->getDSGraph(*A->getParent());
-  else if (const BasicBlock *BB = dyn_cast<BasicBlock>(V))
-    return TD->getDSGraph(*BB->getParent());
-  else if (isa<GlobalValue>(V))
-    return TD->getGlobalsGraph();
-
-  // XXX I know this looks bad, but it works for now
-  for (auto U : V->users()) {
-    return getGraphForValue(U);
-  }
-
-  llvm_unreachable("Unexpected value.");
 }
 
-int DSAWrapper::getOffset(const MemoryLocation *l) {
-  const DSGraph::ScalarMapTy &S = getGraphForValue(l->Ptr)->getScalarMap();
-  DSGraph::ScalarMapTy::const_iterator I = S.find((const Value *)l->Ptr);
-  if (I == S.end())
-    return 0;
-  if (I->second.getNode() && I->second.getNode()->isCollapsedNode())
-    return -1;
-  unsigned offset = I->second.getOffset();
-  assert(offset <= INT_MAX && "Cannot handle large offsets");
-  return (int)offset;
-}
-
-bool DSAWrapper::isMemcpyd(const llvm::DSNode *n) {
-  const llvm::EquivalenceClasses<const llvm::DSNode *> &eqs =
-      nodeEqs->getEquivalenceClasses();
-  const llvm::DSNode *nn = eqs.getLeaderValue(n);
-  for (unsigned i = 0; i < memcpys.size(); i++)
-    if (memcpys[i] == nn)
-      return true;
-  return false;
+void DSAWrapper::collectMemCpyds(llvm::Module &M) {
+  for (auto& f : M) {
+    for (auto& b : f) {
+      for (auto& i : b) {
+        if (auto memcpyInst = dyn_cast<MemCpyInst>(&i)) {
+          auto srcNode = getNode(memcpyInst->getSource());
+          auto destNode = getNode(memcpyInst->getDest());
+          if (srcNode)
+            memCpyds.insert(srcNode);
+          if (destNode)
+            memCpyds.insert(destNode);
+        }
+        if (auto memsetInst = dyn_cast<MemSetInst>(&i)) {
+          auto destNode = getNode(memsetInst->getDest());
+          if (destNode)
+            memCpyds.insert(destNode);
+        }
+      }
+    }
+  }
 }
 
-bool DSAWrapper::isStaticInitd(const llvm::DSNode *n) {
-  const llvm::EquivalenceClasses<const llvm::DSNode *> &eqs =
-      nodeEqs->getEquivalenceClasses();
-  const llvm::DSNode *nn = eqs.getLeaderValue(n);
-  for (unsigned i = 0; i < staticInits.size(); i++)
-    if (staticInits[i] == nn)
-      return true;
-  return false;
+void DSAWrapper::countGlobalRefs() {
+  for (auto& g : DG->globals()) {
+    auto& cellRef = g.second;
+    auto* node = cellRef->getNode();
+    if (node) {
+      if (!globalRefCount.count(node))
```

Comment:
What do we need this for?

Reply:
This is needed because we want to make sure that no more than one global variable refer to the same node otherwise we may end up having two non-aliasing pointers point to the same singleton region.